app/form/SeasQDDiscreteGreenOperator.cpp

```diff
@@ -166,4 +229,444 @@ void SeasQDDiscreteGreenOperator::compute_boundary_traction() {
+  CHKERRTHROW(PetscPrintf(PetscObjectComm((PetscObject)G_),"  status: computed %D / pending %D\n",current_gf_, n_gf_ - current_gf_));
+
+  write_facet_labels_IS(mesh);
+}
+
+void SeasQDDiscreteGreenOperator :: write_facet_labels_IS(LocalSimplexMesh<DomainDimension> const& mesh)
+{
+  PetscViewer v;
+  PetscLogDouble t0,t1;
+  std::size_t bndNo,d;
+  PetscInt nfacets = (PetscInt)base::adapter().fault_map().local_size();
+  PetscInt facet_size = (PetscInt)(DomainDimension);
+  PetscInt *_idx;
+  IS is;
+
+  CHKERRTHROW(PetscCalloc1(nfacets*facet_size,&_idx));
+
+  for (bndNo = 0; bndNo < base::adapter().fault_map().local_size(); ++bndNo) {
+    auto fctNo = base::adapter().fault_map().fctNo(bndNo);
+    auto facets = mesh.facets()[fctNo];
+    for (d=0; d<facet_size; d++) {
+      _idx[bndNo * facet_size + d] = (PetscInt)facets[d];
+    }
+  }
+
+  CHKERRTHROW(ISCreateGeneral(PetscObjectComm((PetscObject)G_),nfacets*facet_size,(const PetscInt*)_idx,PETSC_USE_POINTER,&is));
+
+  CHKERRTHROW(PetscTime(&t0));
+  CHKERRTHROW(PetscViewerBinaryOpen(PetscObjectComm((PetscObject)G_),gf_facet_filename_.c_str(),FILE_MODE_WRITE,&v));
+  CHKERRTHROW(PetscViewerDestroy(&v));
+  CHKERRTHROW(PetscTime(&t1));
+  CHKERRTHROW(PetscPrintf(PetscObjectComm((PetscObject)G_),"write_discrete_greens_operator():facets %1.2e (sec)\n",(double)(t1 - t0)));
+
+  CHKERRTHROW(ISDestroy(&is));
+  CHKERRTHROW(PetscFree(_idx));
+}
+
+IS SeasQDDiscreteGreenOperator :: load_facet_labels_seq_IS(void)
+{
+  PetscViewer v;
+  PetscLogDouble t0,t1;
+  int rank;
+  IS is = NULL;
+
+  MPI_Comm_rank(base::comm(),&rank);
+  if (rank == 0) {
+    CHKERRTHROW(ISCreate(PETSC_COMM_SELF,&is));
+    CHKERRTHROW(PetscTime(&t0));
+    CHKERRTHROW(PetscViewerBinaryOpen(PETSC_COMM_SELF,gf_facet_filename_.c_str(),FILE_MODE_READ,&v));
+    CHKERRTHROW(ISLoad(is,v));
+    CHKERRTHROW(PetscViewerDestroy(&v));
+    CHKERRTHROW(PetscTime(&t1));
+    CHKERRTHROW(PetscPrintf(PETSC_COMM_SELF,"load_discrete_greens_operator():facets %1.2e (sec)\n",(double)(t1 - t0)));
+  }
+  return is;
+}
+
+void SeasQDDiscreteGreenOperator :: create_permutation_redundant_IS(LocalSimplexMesh<DomainDimension> const& mesh, IS is)
+{
+  MPI_Comm comm = base::comm();
+  std::size_t bndNo;
+  PetscInt nfacets = (PetscInt)base::adapter().fault_map().local_size();
+  PetscInt facet_size = (PetscInt)(DomainDimension);
+  PetscInt *_idx = NULL,is_len;
+  PetscInt *fault_map_index = NULL;
+  int rank;
+
+  MPI_Comm_rank(comm,&rank);
+  if (rank == 0) {
+    CHKERRTHROW(ISGetSize(is,&is_len));
+    CHKERRTHROW(ISGetIndices(is,(const PetscInt**)&_idx));
+  }
+
+  MPI_Bcast((void*)&is_len,1,MPIU_INT,0,comm);
+
+  if (!_idx) {
+    CHKERRTHROW(PetscCalloc1(is_len,&_idx));
+  }
+
+  MPI_Bcast((void*)_idx,is_len,MPIU_INT,0,comm);
+
+  CHKERRTHROW(PetscCalloc1((PetscInt)base::adapter().fault_map().local_size(),&fault_map_index));
+
+  for (bndNo = 0; bndNo < base::adapter().fault_map().local_size(); ++bndNo) {
```

Comment:
You could replace this by an O(N) algorithm.

```c++
using map_t = std::unordered_map<Simplex<DomainDimension-1u>, std::size_t, SimplexHash<DomainDimension-1u>>;
auto map = map_t(base::adapter().fault_map().local_size());
std::size_t map_index = 0;
constexpr PetscInt facet_size = (PetscInt)(DomainDimension);
 for (n=0; n<is_len/facet_size; n++) {
     auto plex = Simplex<DomainDimension-1u>{};
      for (PetscInt d=0; d<facet_size; d++) {
        plex[d] = _idx[n*facet_size + d];
      }
      map[plex] = map_index++;
}

  for (bndNo = 0; bndNo < base::adapter().fault_map().local_size(); ++bndNo) {
    auto fctNo = base::adapter().fault_map().fctNo(bndNo);
    auto it = map.find(mesh.facets()[fctNo]);
    if (it == map.end()) {
          throw ...; // Error
    }
    fault_map_index[bndNo] = it->second;
}
```

Reply:
I followed the suggestion